File: Java/DesignPatterns/ChainOfResponsibilityDesignPattern.java
package DesignPatterns;

class Request{
    int priority;
}

abstract class IHandler{
    private IHandler nextHandler;

    IHandler setNextHandler(IHandler handler){
        this.nextHandler = handler;
        return this.nextHandler;
    }
    IHandler getNextHandler(){
        return nextHandler;
    }
    abstract boolean isAllowed(Request request);
    abstract void processRequest(Request request);
}

class Level1Handler extends IHandler{

    @Override
    boolean isAllowed(Request request) {
        if(request.priority < 8)
            return true;
        else
            return false;
    }

    @Override
    void processRequest(Request request) {
        if(isAllowed(request)){
            System.out.println("Processing the request : lvel1 is processing");
        }else{
            getNextHandler().processRequest(request);
        }
    }

}

class Level2Handler extends IHandler{

    @Override
    boolean isAllowed(Request request) {
        if(request.priority < 15)
            return true;
        else
            return false;
    }

    @Override
    void processRequest(Request request) {
        if(isAllowed(request)){
            System.out.println("Processing the request : lvel1 is processing");
        }else{
            getNextHandler().processRequest(request);
        }
       
    }

}

class Level3Handler extends IHandler{

    @Override
    boolean isAllowed(Request request) {
        if(request.priority < 2)
            return true;
        else
            return false;
    }
    void processRequest(Request request) {
        if(isAllowed(request)){
            System.out.println("Processing the request : lvel1 is processing");
        }else{
            getNextHandler().processRequest(request);
        }
    }
}

public class ChainOfResponsibilityDesignPattern {
    public static void main(String[] args) {
        Level1Handler level1 = new Level1Handler();
        Level2Handler level2 = new Level2Handler();
        Level3Handler level3 = new Level3Handler();

        // level1.setNextHandler(level2);
        // level2.setNextHandler(level3);
        
        level1.setNextHandler(level2).setNextHandler(level3);


    }
}

File: Golang/DesignPatterns/ChainOfResponsibilityDesignPattern.go
package main

import "fmt"

type Request struct {
	priority int
}

type Handler interface {
	SetNextHandler(handler Handler) Handler
	GetNextHandler() Handler
	IsAllowed(request Request) bool
	ProcessRequest(request Request)
}

type BaseHandler struct {
	nextHandler Handler
}

func (h *BaseHandler) SetNextHandler(handler Handler) Handler {
	h.nextHandler = handler
	return h.nextHandler
}

func (h *BaseHandler) GetNextHandler() Handler {
	return h.nextHandler
}

type Level1Handler struct {
	BaseHandler
}

func (h *Level1Handler) IsAllowed(request Request) bool {
	return request.priority < 8
}

func (h *Level1Handler) ProcessRequest(request Request) {
	if h.IsAllowed(request) {
		fmt.Println("Processing the request : lvel1 is processing")
	} else {
		h.GetNextHandler().ProcessRequest(request)
	}
}

type Level2Handler struct {
	BaseHandler
}

func (h *Level2Handler) IsAllowed(request Request) bool {
	return request.priority < 15
}

func (h *Level2Handler) ProcessRequest(request Request) {
	if h.IsAllowed(request) {
		fmt.Println("Processing the request : lvel1 is processing")
	} else {
		h.GetNextHandler().ProcessRequest(request)
	}
}

type Level3Handler struct {
	BaseHandler
}

func (h *Level3Handler) IsAllowed(request Request) bool {
	return request.priority < 2
}

func (h *Level3Handler) ProcessRequest(request Request) {
	if h.IsAllowed(request) {
		fmt.Println("Processing the request : lvel1 is processing")
	} else {
		h.GetNextHandler().ProcessRequest(request)
	}
}

func main() {
	level1 := &Level1Handler{}
	level2 := &Level2Handler{}
	level3 := &Level3Handler{}

	level1.SetNextHandler(level2).SetNextHandler(level3)
}
